File: pysample/src/Scheduler.py
import Const
import Action


class Scheduler(object):

    def __init__(self, channel):
        self.channel = channel
        n = channel.get_config((Const.main_config, Const.node_count))
        self.action = [None] + [Action.Action(i + 1, self) for i in xrange(n)]
        self.channel_map = {}

    def get_id(self):
        return self.channel.get_id()

    def on_recv(self, message):
        if message[Const.err_code] != Const.ERR_CODE_NONE:
            return
        {
            Const.CALL_TYPE_PREPARE: self.on_prepare,
            Const.CALL_TYPE_SEND: self.on_send,
            Const.CALL_TYPE_SYS: self.on_sys,
            Const.CALL_TYPE_CHANNEL_BUILD: self.on_channel_build,
            Const.CALL_TYPE_CHANNEL_DESTROY: self.on_destroy
        }[message[Const.call_type]](message)

    def on_prepare(self, message):
        self.action[message[Const.sys_message][Const.target]].on_prepare(message)

    def on_send(self, message):
        self.action[message[Const.sys_message][Const.target]].on_send(message)

    def on_sys(self, message):
        pass

    def on_channel_build(self, message):
        target = message[Const.sys_message][Const.target]
        if message[Const.channel_id] != 0:
            self.action[target].on_succ(message)
            self.channel_map[message[Const.channel_id]] = target
        else:
            state = message[Const.state]
            if state == Const.STATE_NOTICE:
                self.action[target].on_request(message)
            elif state == Const.STATE_REFUSE:
                self.action[target].on_refuse(message)

    def on_destroy(self, message):
        channel_id = message[Const.channel_id]
        target = self.channel_map.get(channel_id, 0)
        if target != 0:
            self.action[target].on_destroy(message)
            del self.channel_map[channel_id]

    def send_channel_build(self, target, state, err_code, channel_type):
        self.do_send({
            Const.call_type: Const.CALL_TYPE_CHANNEL_BUILD,
            Const.state: state,
            Const.sys_message: {
                Const.target: target,
                Const.data: "",
                Const.delay: 0
            },
            Const.err_code: err_code,
            Const.channel_type: channel_type
        }, 0)

    def do_send(self, message, target):
        msg = Const.get_empty_message()
        msg.update(message)
        self.channel.send(msg, target)

File: samplecpp/src/Scheduler.cpp
#include "Scheduler.h"
#include "Const.h"

Scheduler::Scheduler(Channel *channel) {
    this->channel = channel;
    std::vector<std::string> args {Const::main_config, Const::node_count};
    int n = channel->getConfig(args);
    action.resize(n + 1);
    for (int i=1 ; i<=n ; i++) {
        action[i] = new Action(i, this);
    }
    channelMap.clear();
}

int Scheduler::getId() {
    return getChannel()->getId();
}

void Scheduler::onRecv(Json message) {
    if (message[Const::err_code] != Const::ERR_CODE_NONE){
        return ;
    }
    std::string call = message[Const::call_type];
    if (call == Const::CALL_TYPE_PREPARE) onPrepare(message);
    else if (call == Const::CALL_TYPE_SEND) onSend(message);
    else if (call == Const::CALL_TYPE_SYS) onSys(message);
    else if (call == Const::CALL_TYPE_CHANNEL_BUILD) onChannelBuild(message);
    else if (call == Const::CALL_TYPE_CHANNEL_DESTROY) onDestroy(message);
}

void Scheduler::onPrepare(Json message) {
    Print("prepare");
    action[message[Const::sys_message][Const::target]]->onPrepare(message);
}

void Scheduler::onSend(Json message) {
    Print("send")
    action[message[Const::sys_message][Const::target]]->onSend(message);
}

void Scheduler::onChannelBuild(Json message) {
    Print("build")
    int target = message[Const::sys_message][Const::target];
    if (message[Const::channel_id] != 0) {
        Print("succ")
        action[target]->onSucc(message);
        channelMap[message[Const::channel_id]] = target;
    } else {
        int state = message[Const::state];
        if (state == Const::STATE_NOTICE) {
            Print("request")
            action[target]->onRequest(message);
        } else if (state == Const::STATE_REFUSE) {
            Print("refuse")
            action[target]->onRefuse(message);
        }
    }
}

void Scheduler::onDestroy(Json message) {
    Print("destroy");
    int channelId = message[Const::channel_id];
    auto target = channelMap.find(channelId);
    if (target != channelMap.end()) {
        action[target->second]->onDestroy(message);
        channelMap.erase(target);
    }
}

void Scheduler::sendChannelBuild(int target, int state, int errCode, int channelType) {
    Json msg = Const::empty_message;
    msg[Const::call_type] = Const::CALL_TYPE_CHANNEL_BUILD;
    msg[Const::state] = state;
    msg[Const::sys_message][Const::target] = target;
    msg[Const::sys_message][Const::data] = "";
    msg[Const::sys_message][Const::delay] = 0;
    msg[Const::err_code] = errCode;
    msg[Const::channel_type] = channelType;
    doSend(msg, 0);
}

void Scheduler::doSend(Json message, int target) {
    channel->send(message, target);
}

Channel* Scheduler::getChannel() {
    return channel;
}